qiita_db/processing_job.py

```diff
@@ -6,21 +6,291 @@
 # The full license is in the file LICENSE, distributed with this software.
 # -----------------------------------------------------------------------------
 
-from uuid import UUID
+import networkx as nx
+import qiita_db as qdb
+from collections import defaultdict, Iterable
 from datetime import datetime
-from subprocess import Popen, PIPE
-from multiprocessing import Process
-from os.path import join
+from future.utils import viewitems, viewvalues
 from itertools import chain
-from collections import defaultdict, Iterable
 from json import dumps, loads
+from multiprocessing import Process, Queue, Event
+from qiita_core.qiita_settings import qiita_config
+from qiita_db.util import create_nested_path
+from re import search, findall
+from subprocess import Popen, PIPE
 from time import sleep
+from uuid import UUID
+from os.path import join
 
-from future.utils import viewitems, viewvalues
-import networkx as nx
 
-from qiita_core.qiita_settings import qiita_config
-import qiita_db as qdb
+class Watcher(Process):
+    # TODO: Qiita will need a proper mapping of these states to Qiita states
+    # Currently, these strings are being inserted directly into Qiita's status
+    # table. Qiita will be unfamiliar with many of these. We will need at least
+    # one additional job type for 'Held': A job waiting for another to complete
+    # before it can run.
+    #
+    # Note that the main Qiita script instantiates an object of this class in
+    # a separate thread, so it can periodically update the database w/metadata
+    # from Watcher's queue. Qiita's script also calls qdb.complete() so there
+    # are no circular references. TODO: replace w/a REST call.
+
+    # valid Qiita states:
+    #             The current status of the job, one of {'queued', 'running',
+    #             'success', 'error', 'in_construction', 'waiting'}
+
+    # TODO: what to map in_construction to?
+    job_state_map = {'C': 'completed', 'E': 'exiting', 'H': 'held',
+                     'Q': 'queued', 'R': 'running', 'T': 'moving',
+                     'W': 'waiting', 'S': 'suspended'}
+
+    torque_to_qiita_state_map = {'completed': 'completed',
+                                 'exiting': 'running',
+                                 'held': 'queued',
+                                 'queued': 'queued',
+                                 'running': 'running',
+                                 'moving': 'waiting',
+                                 'waiting': 'waiting',
+                                 'suspended': 'waiting',
```

Comment:
Yeah, I'd hate to map torque.suspended to qiita.running, but I think that with respect to Qiita, a process that was suspended in order to be moved, or whatever, would still be thought of as 'running' with respect to Qiita. Perhaps even torque.moving and torque.waiting could be thought of as qiita.running, since in Qiita, waiting explicitly means waiting on other associated jobs to finish.

With respect to reporting current job state to Qiita users in the UI, it'd be great if we had additional states in Qiita, but currently, they're not needed for the backend. I tried not to add new states in Qiita at this point, without looking into the ramifications. What do you think about adding new states?

Reply:
I have added a new state 'DROPPED', which is what is given to the database when say a validator job is dropped because the previous job completed with an error. At this point, should we map this to completed?

Reply:
I guess it should be error too, no? but perhaps dropped as a new state will work ...

Reply:
tentatively keeping it DROPPED